pkg/datapath/linux/node.go

```diff
@@ -1485,13 +1485,17 @@ func (n *linuxNodeHandler) NodeConfigurationChanged(newConfig datapath.LocalNode
 		switch {
 		case !option.Config.EnableL2NeighDiscovery:
 			n.enableNeighDiscovery = false
-		case option.Config.EnableNodePort:
-			mac, err := link.GetHardwareAddr(option.Config.DirectRoutingDevice)
-			if err != nil {
-				return err
+		case option.Config.DirectRoutingDeviceRequired():
+			if option.Config.DirectRoutingDevice != "" {
+				mac, err := link.GetHardwareAddr(option.Config.DirectRoutingDevice)
+				if err != nil {
+					return err
+				}
+				ifaceName = option.Config.DirectRoutingDevice
+				n.enableNeighDiscovery = mac != nil // No need to arping for L2-less devices
+			} else {
+				return fmt.Errorf("direct routing device is required, but not defined")
```

Comment:
No, I haven't observed such a case before. I just defensively added it. I may miss something, but it's critical to make a mistake in the future because we may load a datapath which `DIRECT_ROUTING_DEV_IFINDEX 0` and BPF NodePort or BPF Host Routing enabled. In that case, we'll load the program silently drop the packets, for example, here.

https://github.com/cilium/cilium/blob/master/bpf/lib/fib.h#L20-L63 (redirect to 0 ifindex)

Reply:
Yep, makes sense to me.